docs/architecture/adr-062-p2p-architecture.md

```diff
@@ -0,0 +1,531 @@
+
+Data is exchanged over IO streams created with `Connection.Stream()` using an arbitrary `StreamID`. These implement the standard Go `io.Reader` and `io.Writer` interfaces to read and write bytes. Transports are free to choose how to implement such streams, e.g. by taking advantage of native stream support in the underlying protocol or through multiplexing.
+
+`Connection` and the related `Stream` interfaces are:
+
+```go
+// Connection represents an established connection between two endpoints.
+type Connection interface {
+    // Stream opens a logically distinct IO stream within the connection,
+    // using an arbitrary stream ID. The stream must be created before it
+    // can receive data, and must be closed before the ID can be reused.
+    Stream(StreamID) Stream
+
+    // LocalEndpoint returns the local endpoint for the connection.
+    LocalEndpoint() Endpoint
+
+    // RemoteEndpoint returns the remote endpoint for the connection.
+    RemoteEndpoint() Endpoint
+
+    // PubKey returns the public key of the remote peer.
+    PubKey() crypto.PubKey
+
+    // Close closes the connection.
+    Close() error
+}
+
+// StreamID is an arbitrary stream ID.
+type StreamID uint8
+
+// Stream represents a single logical IO stream within a connection.
+type Stream interface {
+    io.Reader // Read([]byte) (int, error)
+    io.Writer // Write([]byte) (int, error)
+    io.Closer // Close() error
+}
+```
+
+### Peers
+
+Peers are other Tendermint network nodes. Each peer is identified by a unique `PeerID`, and has a set of `PeerAddress` addresses expressed as URLs that they can be reached at. Examples of peer addresses might be e.g.:
+
+* `mconn://b10c@host.domain.com:25567/path`
+* `unix:///var/run/tendermint/peer.sock`
+* `memory:testpeer`
+
+Addresses are resolved into one or more transport endpoints, e.g. by resolving DNS hostnames into IP addresses (which should be refreshed periodically). Peers should always be expressed as address URLs, and never as endpoints which are a lower-level construct.
+
+```go
+// PeerID is a unique peer ID, generally expressed in hex form.
+type PeerID []byte
+
+// PeerAddress is a peer address URL. The User field, if set, gives the
+// hex-encoded remote PeerID, which should be verified with the remote peer's
+// public key as returned by the connection.
+type PeerAddress url.URL
+
+// Resolve resolves a PeerAddress into a set of Endpoints, typically by
+// expanding out any DNS names given in Host to IP addresses. Field mapping:
+//
+//   Scheme → Endpoint.Protocol
+//   Host   → Endpoint.IP
+//   Port   → Endpoint.Port
+//   Path+Query+Fragment,Opaque → Endpoint.Path
+//
+func (a PeerAddress) Resolve(ctx context.Context) []Endpoint { return nil }
```

Comment:
Why does each address resolve to multiple Endpoints?

Reply:
That's how DNS works:

```
$ dig www.netflix.com +short
www.dradis.netflix.com.
www.eu-west-1.internal.dradis.netflix.com.
dualstack.apiproxy-website-nlb-prod-2-b4de62b516adfbbf.elb.eu-west-1.amazonaws.com.
54.155.246.232
18.200.8.190
54.73.148.110
```

We want to get all IPs, and track e.g. connection statistics for them so we can use the best one.